File: androidApp/src/main/kotlin/com/humblesolutions/finai/ui/setup/SetupViewModel.kt
package com.humblesolutions.finai.ui.setup

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.humblesolutions.finai.config.ApiConfig
import com.humblesolutions.finai.config.Supabase
import com.humblesolutions.finai.data.KtorCapabilitiesRepository
import com.humblesolutions.finai.data.KtorFinancialSetupRepository
import com.humblesolutions.finai.data.SupabaseTokenSource
import com.humblesolutions.finai.i18n.Strings
import com.humblesolutions.finai.model.ApiException
import com.humblesolutions.finai.model.Capabilities
import com.humblesolutions.finai.model.FinancialSetup
import com.humblesolutions.finai.repository.CapabilitiesRepository
import com.humblesolutions.finai.repository.FinancialSetupRepository
import com.humblesolutions.finai.usecase.ItemDraft
import com.humblesolutions.finai.usecase.SetupStep
import com.humblesolutions.finai.usecase.SetupWizard
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlin.coroutines.cancellation.CancellationException

/**
 * Drives the financial setup wizard (PRD F1, ticket #17).
 *
 * The repositories are built fresh in [bind] and closed with the model — never
 * singletons (kmp-arch-v2). Which step may be left, and what the notice says,
 * are never decided here: [SetupWizard] answers both, and iOS reads the same
 * rules.
 *
 * **Continue does not wait.** The step slides the moment it is pressed and the
 * save runs behind it, shown only by a small loader, and only while a request is
 * really in flight: a draft the server already holds is not sent again. The last
 * step is the one exception — the gate clears on that save, so Complete waits.
 * Every save still carries the whole draft, so closing the app resumes where it
 * left off and a failed save keeps what was typed.
 */
class SetupViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(SetupUiState())
    val uiState: StateFlow<SetupUiState> = _uiState.asStateFlow()

    private var repository: FinancialSetupRepository? = null
    private var capabilities: CapabilitiesRepository? = null

    /** Whose wizard this is holding. Null until the first bind. */
    private var boundTo: String? = null

    /**
     * Bumped every time the wizard is handed to someone new. A request records
     * the value it set out under, and its reply is applied only if that is still
     * the value — so a slow answer for the last user can never land in the next
     * one's wizard (kmp-arch-v2 → guard async results against staleness).
     */
    private var generation = 0

    /**
     * What the server holds, as far as this model knows — last loaded or last
     * saved. A draft equal to it is not sent again, which is what keeps moving
     * back and forth through the steps free of loaders.
     */
    private var lastSaved: FinancialSetup? = null

    /** The save loop, while one is running. */
    private var saver: Job? = null

    /**
     * Binds to [userId], and stays bound while that user is the one signed in.
     *
     * A `ViewModel` rather than something remembered by the screen, so that a
     * rotation recreates the composition and finds this again — with the figure
     * being typed still in it. That longevity is why the id is checked: this
     * model outlives a sign-out, so binding without asking who wants it would
     * hand the next person the previous one's draft.
     */
    fun bind(userId: String, logging: Boolean) = bind(userId) {
        Supabase.clientOrNull()?.let { client ->
            val tokens = SupabaseTokenSource(client)
            SetupRepositories(
                setup = KtorFinancialSetupRepository(ApiConfig.BASE_URL, tokens, logging),
                capabilities = KtorCapabilitiesRepository(ApiConfig.BASE_URL, tokens, logging),
            )
        }
    }

    /**
     * The whole of [bind], with where the repositories come from left open, so
     * a test drives exactly the path the app does rather than a piece of it.
     */
    internal fun bind(userId: String, build: () -> SetupRepositories?) {
        if (userId.isBlank()) return refuseUnknownUser()
        if (userId == boundTo) return
        releaseTo(userId)
        val built = build() ?: return
        repository = built.setup
        capabilities = built.capabilities
        load()
    }

    /**
     * Drops the clients that carried the last user's token, and everything they
     * typed, so nothing crosses from one account into the next — including any
     * reply still on its way for them.
     */
    private fun releaseTo(userId: String?) {
        closeClients()
        saver?.cancel()
        saver = null
        lastSaved = null
        boundTo = userId
        generation++
        _uiState.value = SetupUiState()
    }

    /**
     * No id to bind to. Nothing of the last user may stay on screen, and a
     * loader that never stops is not an answer, so this says so instead.
     */
    private fun refuseUnknownUser() {
        releaseTo(null)
        _uiState.value = SetupUiState(loading = false, errorKey = Strings.error_unexpected)
    }

    override fun onCleared() = closeClients()

    private fun closeClients() {
        repository?.close()
        repository = null
        capabilities?.close()
        capabilities = null
    }

    /** What is already saved decides where the wizard opens. */
    fun load() {
        val repository = repository ?: return
        val capabilities = capabilities
        val started = generation
        _uiState.update { it.copy(loading = true, errorKey = null) }
        viewModelScope.launch {
            try {
                val saved = repository.get()
                val resume = SetupWizard.resumeAt(saved)
                // The currency and the language figures are written in come
                // from capabilities (ticket #17). The setup response names a
                // currency too, which stands in when capabilities cannot be
                // had: what this screen must not do is guess from the device.
                val payload = capabilitiesOrNull(capabilities)
                if (started != generation) return@launch
                _uiState.update {
                    it.copy(
                        loading = false,
                        currency = payload?.currency?.ifBlank { null } ?: saved.currency,
                        locale = payload?.locale.orEmpty(),
                        draft = SetupWizard.draftFrom(saved),
                        step = resume,
                        touched = false,
                    )
                }
                // What was just loaded is what the server holds, so walking
                // through the steps without changing anything sends nothing.
                lastSaved = payloadOf(_uiState.value)
            } catch (e: CancellationException) {
                throw e
            } catch (e: ApiException) {
                if (started != generation) return@launch
                _uiState.update { it.copy(loading = false, errorKey = e.messageKey) }
            }
        }
    }

    /**
     * How the wizard is asked for capabilities: it decorates the screen, so a
     * failure to fetch it is not a failure of the screen.
     */
    private suspend fun capabilitiesOrNull(capabilities: CapabilitiesRepository?): Capabilities? = try {
        capabilities?.fetch()
    } catch (e: CancellationException) {
        throw e
    } catch (e: ApiException) {
        null
    }

    fun onIncomeChange(value: String) = _uiState.update {
        it.copy(draft = it.draft.copy(income = value), errorKey = null, touched = true)
    }

    fun onExpenseChange(value: String) = _uiState.update {
        it.copy(draft = it.draft.copy(monthlyExpense = value), errorKey = null, touched = true)
    }

    /** Back a step, or nothing to do on the first one. */
    fun back() = _uiState.update {
        when (it.step) {
            SetupStep.INCOME -> it
            SetupStep.EXPENSES -> it.copy(step = SetupStep.INCOME, errorKey = null, touched = false)
            SetupStep.PORTFOLIO -> it.copy(step = SetupStep.EXPENSES, errorKey = null, touched = false)
        }
    }

    /**
     * Moves on at once and saves behind the move. The last step instead
     * finishes, which waits for its save: the gate clears on it.
     */
    fun continueStep(onFinished: () -> Unit) {
        val state = _uiState.value
        if (!state.canContinue) return
        if (state.step.isLast) return finish(onFinished)
        _uiState.update { it.copy(step = next(it.step), errorKey = null, touched = false) }
        persist()
    }

    /**
     * Past the last step without filling it in: an empty list is the record
     * (#29). Offered only once both mandatory figures are in, so it always goes
     * through.
     */
    fun skip(onFinished: () -> Unit) {
        val state = _uiState.value
        if (!state.canSkip) return
        _uiState.update {
            it.copy(draft = it.draft.copy(debts = emptyList(), investments = emptyList()), touched = false)
        }
        finish(onFinished)
    }

    /** The user swiped. Any step may be looked at; Continue is what stays gated. */
    fun goTo(step: SetupStep) = _uiState.update {
        if (it.step == step) it else it.copy(step = step, errorKey = null, touched = false)
    }

    /** Cancel: the wizard says setup is required, keeping what was typed. */
    fun cancel() = _uiState.update {
        it.copy(cancelled = true, editing = null, rows = emptyList(), errorKey = null)
    }

    /** Back from the required-setup screen, to step 1. */
    fun resume() = _uiState.update {
        it.copy(cancelled = false, step = SetupStep.INCOME, errorKey = null, touched = false)
    }

    // ── The itemised lists ──────────────────────────────────────────────

    fun openList(list: ItemList) = _uiState.update {
        it.copy(editing = list, rows = it.itemsOf(list).ifEmpty { listOf(ItemDraft()) }, errorKey = null)
    }

    fun onRowChange(index: Int, row: ItemDraft) = _uiState.update {
        it.copy(rows = it.rows.mapIndexed { i, existing -> if (i == index) row else existing }, errorKey = null)
    }

    fun addRow() = _uiState.update { it.copy(rows = it.rows + ItemDraft()) }

    fun removeRow(index: Int) = _uiState.update {
        val remaining = it.rows.filterIndexed { i, _ -> i != index }
        it.copy(rows = remaining)
    }

    /** Keeps the rows and saves them behind the step, so a list survives the app closing. */
    fun keepRows() {
        val state = _uiState.value
        val list = state.editing ?: return
        if (!state.canKeepRows) return
        val kept = state.rows.filterNot { it.name.isBlank() && it.amount.isBlank() }
        val draft = when (list) {
            ItemList.OBLIGATIONS -> state.draft.copy(obligations = kept)
            ItemList.DEBTS -> state.draft.copy(debts = kept)
            ItemList.INVESTMENTS -> state.draft.copy(investments = kept)
        }
        _uiState.update { it.copy(draft = draft, editing = null, rows = emptyList()) }
        persist()
    }

    /** Leaves the list as it was before it was opened. */
    fun discardRows() = _uiState.update { it.copy(editing = null, rows = emptyList(), errorKey = null) }

    // ── Saving ──────────────────────────────────────────────────────────

    /**
     * Finishes: waits for the draft on screen to be stored, then hands back to
     * onboarding, which reloads `/me` — the gate clears once income and expense
     * are in. A failed save stays here with its message and the figures intact.
     */
    private fun finish(onFinished: () -> Unit) {
        val started = generation
        _uiState.update { it.copy(busy = true, errorKey = null) }
        viewModelScope.launch {
            persist()?.join()
            if (started != generation) return@launch
            val stored = lastSaved != null && lastSaved == payloadOf(_uiState.value)
            _uiState.update { it.copy(busy = false) }
            if (stored) onFinished()
        }
    }

    /**
     * Sends the draft as it is now, unless the server already holds exactly
     * that. One loop at a time: a call while it runs returns the running loop,
     * which reads the draft afresh before stopping, so a change made mid-save
     * is sent right after rather than racing it.
     */
    private fun persist(): Job? {
        val repository = repository ?: return null
        saver?.takeIf { it.isActive }?.let { return it }
        val started = generation
        return viewModelScope.launch {
            try {
                while (true) {
                    val payload = payloadOf(_uiState.value)
                    if (payload == lastSaved) break
                    _uiState.update { it.copy(syncing = true, errorKey = null) }
                    repository.save(payload)
                    if (started != generation) return@launch
                    lastSaved = payload
                }
                _uiState.update { it.copy(syncing = false) }
            } catch (e: CancellationException) {
                throw e
            } catch (e: ApiException) {
                if (started != generation) return@launch
                // The draft is untouched: a failed save must never lose figures.
                _uiState.update { it.copy(syncing = false, errorKey = e.messageKey) }
            } catch (e: Exception) {
                if (started != generation) return@launch
                _uiState.update { it.copy(syncing = false, errorKey = Strings.error_unexpected) }
            }
        }.also { saver = it }
    }

    private fun payloadOf(state: SetupUiState): FinancialSetup =
        SetupWizard.payload(state.draft, state.currency, state.fractionDigits)

    private fun next(step: SetupStep): SetupStep = when (step) {
        SetupStep.INCOME -> SetupStep.EXPENSES
        SetupStep.EXPENSES, SetupStep.PORTFOLIO -> SetupStep.PORTFOLIO
    }
}

/** The two clients the wizard needs, built together for one signed-in user. */
internal class SetupRepositories(
    val setup: FinancialSetupRepository,
    val capabilities: CapabilitiesRepository,
)

File: iosApp/iosApp/viewmodel/SetupViewModel.swift
import Foundation
import SharedLogic

/// Drives the financial setup wizard (PRD F1, ticket #17).
///
/// The repositories are built fresh in `bind()` and closed in `unbind()` — never
/// singletons (kmp-arch-v2). Which step may be left, and what the notice says,
/// are never decided here: `SetupWizard` answers both, and Android reads the
/// same rules.
///
/// **Continue does not wait.** The step slides the moment it is pressed and the
/// save runs behind it, shown only by a small loader, and only while a request
/// is really in flight: a draft the server already holds is not sent again. The
/// last step is the one exception — the gate clears on that save, so Complete
/// waits. Every save still carries the whole draft, so closing the app resumes
/// where it left off and a failed save keeps what was typed.
@MainActor
final class SetupViewModel: ObservableObject {

    /// Which itemised list is open, if any.
    enum ItemList { case obligations, debts, investments }

    /// The step on screen. Any step can be swiped to; Continue is what is gated.
    @Published private(set) var step: SetupStep = .income
    @Published private(set) var draft = SetupDraft(income: "", monthlyExpense: "", obligations: [], debts: [], investments: [])
    @Published private(set) var currency = ""
    /// The language figures are formatted in, from capabilities — never the
    /// device's. Blank until it has loaded, which formats as English.
    @Published private(set) var locale = ""
    /// The first load, while the app's coin loader covers the screen.
    @Published private(set) var loading = true
    /// Finishing: the save the gate clears on is under way.
    @Published private(set) var busy = false
    /// A save running behind the steps. Only the small loader says so.
    @Published private(set) var syncing = false
    @Published private(set) var errorKey: String?
    @Published private(set) var editing: ItemList?
    /// The rows being edited, kept apart until they are kept or dropped.
    @Published var rows: [ItemDraft] = []
    /// Whether this step's figure has been edited since the step was shown.
    @Published private(set) var touched = false
    /// The user cancelled setup: the wizard says it is required instead of a step.
    @Published private(set) var cancelled = false

    private var repository: FinancialSetupRepository?
    private var capabilities: CapabilitiesRepository?
    /// Bumped on every unbind. A request records the value it set out under and
    /// applies its reply only if that is still the value, so a slow answer for
    /// the last session cannot land in the next one (kmp-arch-v2).
    private var generation = 0
    /// What the server holds as far as this model knows — last loaded or last
    /// saved. A draft equal to it is not sent again.
    private var lastSaved: FinancialSetup?
    /// The save loop, while one is running.
    private var saver: Task<Void, Never>?

    #if DEBUG
    private let logging = true
    #else
    private let logging = false
    #endif

    // MARK: - Derived

    var fractionDigits: Int32 { Money.shared.fractionDigits(currency: currency) }
    var symbol: String { Money.shared.symbol(currency: currency) }

    /// Why Continue on this step cannot go ahead, or nil — shared with the notice.
    var block: SetupBlock? {
        SetupWizard.shared.blockingReason(step: step, draft: draft, fractionDigits: fractionDigits)
    }

    /// Continue waits for every mandatory figure up to this step, and for a finish.
    var canContinue: Bool { !busy && block == nil }

    /// What the notice under Continue says — the same reason, with the shared timing.
    var notice: SetupBlock? {
        SetupWizard.shared.notice(step: step, draft: draft, fractionDigits: fractionDigits, touched: touched)
    }

    /// Skip is drawn only on the step that is optional in full…
    var showsSkip: Bool { step.isOptional }

    /// …and goes through only once both mandatory figures are in.
    var canSkip: Bool {
        !busy && SetupWizard.shared.canSkip(step: step, draft: draft, fractionDigits: fractionDigits)
    }

    /// The small loader: a save under way. Never the coin.
    var showsSaving: Bool { syncing || busy }

    var canKeepRows: Bool { !busy && rows.indices.allSatisfy { rowBlock(at: $0) == nil } }

    /// What a list adds up to, written for reading, or nil while it is empty.
    func total(of list: ItemList) -> String? {
        guard let sum = SetupWizard.shared.total(
            items: items(of: list),
            fractionDigits: fractionDigits,
            debt: list == .debts
        ) else { return nil }
        let written = Money.shared.format(amount: sum, currency: currency, locale: locale)
        return written.isEmpty ? nil : written
    }

    func rowBlock(at index: Int) -> SetupBlock? {
        guard index < rows.count else { return nil }
        return SetupWizard.shared.itemBlock(
            item: rows[index],
            fractionDigits: fractionDigits,
            debt: editing == .debts
        )
    }

    func items(of list: ItemList) -> [ItemDraft] {
        switch list {
        case .obligations: return draft.obligations
        case .debts: return draft.debts
        case .investments: return draft.investments
        }
    }

    // MARK: - Lifecycle

    func bind() {
        guard repository == nil, let client = Supabase.shared.clientOrNull() else { return }
        let tokens = SupabaseTokenSource(client: client)
        repository = KtorFinancialSetupRepository(
            baseUrl: ApiConfig.shared.BASE_URL,
            tokens: tokens,
            logging: logging
        )
        capabilities = KtorCapabilitiesRepository(
            baseUrl: ApiConfig.shared.BASE_URL,
            tokens: tokens,
            logging: logging
        )
        load()
    }

    func unbind() {
        repository?.close()
        repository = nil
        capabilities?.close()
        capabilities = nil
        saver?.cancel()
        saver = nil
        lastSaved = nil
        // Any reply still on its way belongs to the session being left.
        generation += 1
        // Nothing of this user survives into the next one's session. The model
        // itself outlives a sign-out, so leaving the draft here would show one
        // person's figures to whoever signs in next if their load failed.
        draft = SetupDraft(income: "", monthlyExpense: "", obligations: [], debts: [], investments: [])
        currency = ""
        locale = ""
        step = .income
        editing = nil
        rows = []
        errorKey = nil
        touched = false
        cancelled = false
        busy = false
        syncing = false
        loading = true
    }

    /// What is already saved decides where the wizard opens.
    func load() {
        guard let repository else { return }
        let capabilities = self.capabilities
        let started = generation
        loading = true
        errorKey = nil
        Task { [weak self] in
            guard let self else { return }
            do {
                let saved = try await repository.get()
                // The currency and the language figures are written in come
                // from capabilities (ticket #17). The setup response names a
                // currency too, which stands in when capabilities cannot be
                // had: what this screen must not do is guess from the device.
                let payload = await Self.capabilitiesOrNil(capabilities)
                guard started == self.generation else { return }
                self.currency = payload.map { $0.currency.isEmpty ? saved.currency : $0.currency }
                    ?? saved.currency
                self.locale = payload?.locale ?? ""
                self.draft = SetupWizard.shared.draftFrom(setup: saved)
                self.step = SetupWizard.shared.resumeAt(setup: saved)
                self.touched = false
                // What was just loaded is what the server holds, so walking
                // through the steps without changing anything sends nothing.
                self.lastSaved = self.payload
            } catch {
                guard started == self.generation else { return }
                self.errorKey = Self.messageKey(error)
            }
            self.loading = false
        }
    }

    /// How the wizard asks for capabilities: it decorates the screen, so a
    /// failure to fetch it is not a failure of the screen.
    private static func capabilitiesOrNil(_ capabilities: CapabilitiesRepository?) async -> Capabilities? {
        guard let capabilities else { return nil }
        return try? await capabilities.fetch()
    }

    // MARK: - The figures

    func setIncome(_ value: String) {
        draft = draft.doCopy(
            income: value, monthlyExpense: draft.monthlyExpense,
            obligations: draft.obligations, debts: draft.debts, investments: draft.investments
        )
        errorKey = nil
        touched = true
    }

    func setExpense(_ value: String) {
        draft = draft.doCopy(
            income: draft.income, monthlyExpense: value,
            obligations: draft.obligations, debts: draft.debts, investments: draft.investments
        )
        errorKey = nil
        touched = true
    }

    func back() {
        errorKey = nil
        touched = false
        switch step {
        case .expenses: step = .income
        case .portfolio: step = .expenses
        default: break
        }
    }

    /// Moves on at once and saves behind the move. The last step instead
    /// finishes, which waits for its save: the gate clears on it.
    func continueStep(onFinished: @escaping () -> Void) {
        guard canContinue else { return }
        if step.isLast { return finish(onFinished) }
        step = step == .income ? .expenses : .portfolio
        touched = false
        errorKey = nil
        persist()
    }

    /// Past the last step without filling it in: an empty list is the record
    /// (#29). Offered only once both mandatory figures are in.
    func skip(onFinished: @escaping () -> Void) {
        guard canSkip else { return }
        draft = draft.doCopy(
            income: draft.income, monthlyExpense: draft.monthlyExpense,
            obligations: draft.obligations, debts: [], investments: []
        )
        touched = false
        finish(onFinished)
    }

    /// The user swiped. Any step may be looked at; Continue is what stays gated.
    func goTo(_ destination: SetupStep) {
        guard destination != step else { return }
        step = destination
        errorKey = nil
        touched = false
    }

    /// Cancel: the wizard says setup is required, keeping what was typed.
    func cancel() {
        cancelled = true
        editing = nil
        rows = []
        errorKey = nil
    }

    /// Back from the required-setup screen, to step 1.
    func resume() {
        cancelled = false
        step = .income
        errorKey = nil
        touched = false
    }

    // MARK: - The itemised lists

    func openList(_ list: ItemList) {
        editing = list
        let existing = items(of: list)
        rows = existing.isEmpty ? [ItemDraft(name: "", amount: "", minimumPayment: "", interestRatePercent: "")] : existing
        errorKey = nil
    }

    func addRow() {
        rows.append(ItemDraft(name: "", amount: "", minimumPayment: "", interestRatePercent: ""))
    }

    func removeRow(at index: Int) {
        guard rows.indices.contains(index) else { return }
        rows.remove(at: index)
    }

    /// Keeps the rows and saves them behind the step, so a list survives the app closing.
    func keepRows() {
        guard let list = editing, canKeepRows else { return }
        let kept = rows.filter { !($0.name.isEmpty && $0.amount.isEmpty) }
        switch list {
        case .obligations:
            draft = draft.doCopy(
                income: draft.income, monthlyExpense: draft.monthlyExpense,
                obligations: kept, debts: draft.debts, investments: draft.investments
            )
        case .debts:
            draft = draft.doCopy(
                income: draft.income, monthlyExpense: draft.monthlyExpense,
                obligations: draft.obligations, debts: kept, investments: draft.investments
            )
        case .investments:
            draft = draft.doCopy(
                income: draft.income, monthlyExpense: draft.monthlyExpense,
                obligations: draft.obligations, debts: draft.debts, investments: kept
            )
        }
        editing = nil
        rows = []
        persist()
    }

    /// Leaves the list as it was before it was opened.
    func discardRows() {
        editing = nil
        rows = []
        errorKey = nil
    }

    // MARK: - Saving

    /// Finishes: waits for the draft on screen to be stored, then hands back to
    /// onboarding, which reloads `/me`. A failed save stays here with its message
    /// and the figures intact.
    private func finish(_ onFinished: @escaping () -> Void) {
        let started = generation
        busy = true
        errorKey = nil
        let saving = persist()
        Task { [weak self] in
            await saving?.value
            guard let self, started == self.generation else { return }
            self.busy = false
            if let last = self.lastSaved, last == self.payload { onFinished() }
        }
    }

    /// Sends the draft as it is now, unless the server already holds exactly
    /// that. One loop at a time: a call while it runs returns the running loop,
    /// which reads the draft afresh before stopping, so a change made mid-save is
    /// sent right after rather than racing it.
    @discardableResult
    private func persist() -> Task<Void, Never>? {
        guard let repository else { return nil }
        if let saver { return saver }
        let started = generation
        let task = Task { [weak self] in
            guard let self else { return }
            defer { if started == self.generation { self.saver = nil } }
            do {
                while true {
                    let payload = self.payload
                    if let last = self.lastSaved, last == payload { break }
                    self.syncing = true
                    self.errorKey = nil
                    _ = try await repository.save(setup: payload)
                    guard started == self.generation else { return }
                    self.lastSaved = payload
                }
                self.syncing = false
            } catch {
                guard started == self.generation else { return }
                // The draft is untouched: a failed save must never lose figures.
                self.syncing = false
                self.errorKey = Self.messageKey(error)
            }
        }
        saver = task
        return task
    }

    private var payload: FinancialSetup {
        SetupWizard.shared.payload(draft: draft, currency: currency, fractionDigits: fractionDigits)
    }

    private static func messageKey(_ error: Error) -> String {
        ((error as NSError).userInfo["KotlinException"] as? ApiException)?.messageKey
            ?? Strings.shared.error_unexpected
    }
}
